src/generic/usb_cdc.c

```diff
@@ -358,21 +358,34 @@ usb_req_get_descriptor(struct usb_ctrlrequest *req)
 {
     if (req->bRequestType != USB_DIR_IN)
         goto fail;
+    void *desc = NULL;
+    uint_fast8_t size;
     uint_fast8_t i;
     for (i=0; i<ARRAY_SIZE(cdc_descriptors); i++) {
         const struct descriptor_s *d = &cdc_descriptors[i];
         if (READP(d->wValue) == req->wValue
             && READP(d->wIndex) == req->wIndex) {
-            uint_fast8_t size = READP(d->size);
-            uint_fast8_t flags = NEED_PROGMEM ? UX_SEND_PROGMEM : UX_SEND;
-            if (size > req->wLength)
-                size = req->wLength;
-            else if (size < req->wLength)
-                flags |= UX_SEND_ZLP;
-            usb_do_xfer((void*)READP(d->desc), size, flags);
-            return;
+            size = READP(d->size);
+            desc = (void*)READP(d->desc);
         }
     }
+    if (CONFIG_USB_SERIAL_NUMBER_CHIPID
+        && req->wValue == ((USB_DT_STRING<<8) | USB_STR_ID_SERIAL)
+        && req->wIndex == USB_LANGID_ENGLISH_US) {
+            struct usb_string_descriptor *usbserial_serialid;
+            usbserial_serialid = usbserial_get_serialid();
+            size = usbserial_serialid->bLength;
```

Comment:
I know this is directly contradicting my previous comments, but it occurred to me another way to do this would be to introduce two new platform specific calls: `uint_fast8_t usbserial_serialid_length(void)` and `void usbserial_fill_serialid(uint16_t *data)`.  Then the generic code could do:
```
    size = get_data_length();
    struct {
        struct usb_string_descriptor desc;
        uint16_t data[size];
    } ldesc;
    ldesc.desc.bLength = sizeof(ldesc.desc) + 2 * size;
    ldesc.desc.bDescriptorType = USB_DT_STRING;
    usbserial_fill_serialid(ldesc.data);
```

The main advantage of this approach is that it is not necessary to allocate global memory for the serial id.  Which may be useful on some platforms starved for ram.

I can make this change though, as I don't want you to have to go through multiple revisions.